From: =?utf-8?q?Florian_M=C3=BCllner?= <fmuellner@gnome.org>
Date: Thu, 7 Sep 2023 17:59:03 +0200
Subject: screenshot: Do not wrongly enable window button

The window button is disabled when
 - there are no windows
 - we are in screen-recording mode
 - the session mode doesn't allow windows

However the last condition is only taken into account when
opening the dialog, but not when switching from recording-
to screenshot mode.

Address this by updating the button's sensitivity in a separate
function, so the different conditions are considered consistently.

(cherry picked from commit 521525948eed85cc27c0796a0b9569d161df81ba)

Bug: https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/6990
CVE: CVE-2023-43090
Origin: https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/2944
Applied-upstream: 43.9, commit:9d81bbb8b60aa11082a8e7017995e0bc4745e506
Bug-Debian: https://bugs.debian.org/1052067
---
 js/ui/screenshot.js | 19 ++++++++++++-------
 1 file changed, 12 insertions(+), 7 deletions(-)

diff --git a/js/ui/screenshot.js b/js/ui/screenshot.js
index 4d9e42e..84830b2 100644
--- a/js/ui/screenshot.js
+++ b/js/ui/screenshot.js
@@ -1370,6 +1370,16 @@ var ScreenshotUI = GObject.registerClass({
         this._castButton.reactive = Main.sessionMode.allowScreencast;
     }
 
+    _syncWindowButtonSensitivity() {
+        const windows =
+            this._windowSelectors.flatMap(selector => selector.windows());
+
+        this._windowButton.reactive =
+            Main.sessionMode.hasWindows &&
+            windows.length > 0 &&
+            !this._castButton.checked;
+    }
+
     _refreshButtonLayout() {
         const buttonLayout = Meta.prefs_get_button_layout();
 
@@ -1486,10 +1496,7 @@ var ScreenshotUI = GObject.registerClass({
                 });
             }
 
-            this._windowButton.reactive =
-                Main.sessionMode.hasWindows &&
-                windows.length > 0 &&
-                !this._castButton.checked;
+            this._syncWindowButtonSensitivity();
             if (!this._windowButton.reactive)
                 this._selectionButton.checked = true;
 
@@ -1732,9 +1739,7 @@ var ScreenshotUI = GObject.registerClass({
 
             this._captureButton.remove_style_pseudo_class('cast');
 
-            const windows =
-                this._windowSelectors.flatMap(selector => selector.windows());
-            this._windowButton.reactive = windows.length > 0;
+            this._syncWindowButtonSensitivity();
         }
     }
 
